server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -773,25 +794,32 @@ private String validateAdd(DiscoveryNode node) {
             return null;
         }
 
+        private static <T extends Comparable<T>> T min(T t1, T t2) {
+            assert t2 != null;
+            return t1 == null || t1.compareTo(t2) > 0 ? t2 : t1;
+        }
+
+        private static <T extends Comparable<T>> T max(T t1, T t2) {
+            assert t2 != null;
+            return t1 == null || t1.compareTo(t2) < 0 ? t2 : t1;
+        }
+
         public DiscoveryNodes build() {
             Version minNodeVersion = null;
             Version maxNodeVersion = null;
             Version minNonClientNodeVersion = null;
-            Version maxNonClientNodeVersion = null;
+            IndexVersion maxDataNodeCompatibleIndexVersion = null;
+            IndexVersion minSupportedIndexVersion = null;
             for (Map.Entry<String, DiscoveryNode> nodeEntry : nodes.entrySet()) {
                 DiscoveryNode discoNode = nodeEntry.getValue();
-                final Version version = discoNode.getVersion();
+                Version version = discoNode.getVersion();
                 if (discoNode.canContainData() || discoNode.isMasterNode()) {
-                    if (minNonClientNodeVersion == null) {
-                        minNonClientNodeVersion = version;
-                        maxNonClientNodeVersion = version;
-                    } else {
-                        minNonClientNodeVersion = Version.min(minNonClientNodeVersion, version);
-                        maxNonClientNodeVersion = Version.max(maxNonClientNodeVersion, version);
-                    }
+                    minNonClientNodeVersion = min(minNonClientNodeVersion, version);
+                    maxDataNodeCompatibleIndexVersion = min(maxDataNodeCompatibleIndexVersion, discoNode.getMaxIndexVersion());
```

Comment:
The logic for all of these versions is difficult to understand based on names alone. Can you please add a comment before we figure out all these versions explaining exactly what we are trying to find and the pre-post conditions (eg which can end end up null)?

Reply:
Added a comment